clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/BlobSnapshotManager.java

```diff
@@ -22,66 +25,142 @@
+  // concurrent user count
+  // Example: <topicName, <partitionId, concurrentUsersAmount>>
+  private Map<String, Map<Integer, AtomicLong>> concurrentSnapshotUsers;
+  // A map to keep track of the snapshot timestamps for a particular topic and partition, use to track snapshot
+  // staleness
+  // Example: <topicName, <partitionId, timestamp>>
+  private Map<String, Map<Integer, Long>> snapshotTimestamps;
+  // A map to keep track the snapshot respective offset record for a particular topic and partition, use to keep
+  // snapshot/offset consistency
+  // Example: <topicName, <partitionId, offset>>
+  private Map<String, Map<Integer, BlobTransferPartitionMetadata>> snapshotMetadataRecords;
+
   private final ReadOnlyStoreRepository readOnlyStoreRepository;
+  private final StorageEngineRepository storageEngineRepository;
   private final Lock lock = new ReentrantLock();
-  private static final Logger LOGGER = LogManager.getLogger(BlobSnapshotManager.class);
 
   /**
    * Constructor for the BlobSnapshotManager
    */
   public BlobSnapshotManager(
-      String basePath,
-      long snapshotRetentionTime,
-      ReadOnlyStoreRepository readOnlyStoreRepository) {
-    this.basePath = basePath;
-    this.snapshotRetentionTime = snapshotRetentionTime;
+      ReadOnlyStoreRepository readOnlyStoreRepository,
+      StorageEngineRepository storageEngineRepository) {
     this.readOnlyStoreRepository = readOnlyStoreRepository;
-    this.snapShotTimestamps = new VeniceConcurrentHashMap<>();
+    this.storageEngineRepository = storageEngineRepository;
+
     this.concurrentSnapshotUsers = new VeniceConcurrentHashMap<>();
+    this.snapshotTimestamps = new VeniceConcurrentHashMap<>();
+    this.snapshotMetadataRecords = new VeniceConcurrentHashMap<>();
   }
 
   /**
-   * Checks if the snapshot is stale, if it is and no one is using it, it updates the snapshot,
-   * otherwise it increases the count of people using the snapshot
+   * Recreate a snapshot for a hybrid store and return true if the snapshot is successfully created
+   * If the snapshot is being used by some other host, throw an exception
+   * @param payload the blob transfer payload which carries the topic name, store name and partition id
+   * @param blobTransferPartitionMetadata the blob transfer partition metadata which carries the snapshot offset
    */
-  public void maybeUpdateHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
-    if (rocksDB == null || topicName == null) {
-      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
-    }
-    if (!isStoreHybrid(topicName)) {
-      LOGGER.warn("Store {} is not hybrid, skipping snapshot update", topicName);
-      return;
-    }
-    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
-    snapShotTimestamps.putIfAbsent(topicName, new VeniceConcurrentHashMap<>());
+  public void recreateSnapshotForHybrid(
```

Comment:
If we make this method also work for batch, I am thinking we can do something like this
```
BlobTransferPartitionMetadata  createSnapshot(BlobTransferPayload payload) {
  if batch:
       generate the metadata
       return it
  else:
     // hybrid
     check if we should throttle the request due to many concurrent users
     check whether we have snapshot & metadata available within the staleness threshold
     if yes:
        get the metadata from map
        return metadata
    else:
       generate the metadata
       generate the snapshot
       return metadata
}
```
You can further streamline the logic if needed